orderbook/src/routes/session.rs

```diff
@@ -0,0 +1,137 @@
+use std::time::Duration;
+use std::time::Instant;
+
+use crate::routes::websocket;
+use actix::prelude::*;
+use actix_web_actors::ws;
+
+/// How often heartbeat pings are sent
+const HEARTBEAT_INTERVAL: Duration = Duration::from_secs(5);
+
+/// How long before lack of client response causes a timeout
+const CLIENT_TIMEOUT: Duration = Duration::from_secs(10);
+
+#[derive(Debug)]
+pub struct WsPriceFeedSession {
+    /// unique session id
+    pub id: usize,
+
+    /// Client must send ping at least once per 10 seconds (CLIENT_TIMEOUT),
+    /// otherwise we drop connection.
+    pub hb: Instant,
+
+    /// Pricefeed server
+    pub addr: Addr<websocket::PriceFeed>,
+}
+
+impl WsPriceFeedSession {
+    /// helper method that sends ping to client every 5 seconds (HEARTBEAT_INTERVAL).
+    ///
+    /// also this method checks heartbeats from client
+    fn hb(&self, ctx: &mut ws::WebsocketContext<Self>) {
+        ctx.run_interval(HEARTBEAT_INTERVAL, |act, ctx| {
+            // check client heartbeats
+            if Instant::now().duration_since(act.hb) > CLIENT_TIMEOUT {
+                // heartbeat timed out
+                tracing::warn!(
+                    client_id = act.id,
+                    "Websocket Client heartbeat failed, disconnecting!"
+                );
+
+                // notify price feed server
+                act.addr.do_send(websocket::Disconnect { id: act.id });
+
+                // stop actor
+                ctx.stop();
+
+                // don't try to send a ping
+                return;
+            }
+
+            ctx.ping(b"");
+        });
+    }
+}
+
+impl Actor for WsPriceFeedSession {
+    type Context = ws::WebsocketContext<Self>;
+
+    /// Method is called on actor start.
+    /// We register ws session with PriceFeedServer
+    fn started(&mut self, ctx: &mut Self::Context) {
+        // we'll start heartbeat process on session start.
+        self.hb(ctx);
+
+        // register self in server. `AsyncContext::wait` register
+        // future within context, but context waits until this future resolves
+        // before processing any other events.
+        // HttpContext::state() is instance of SessionState, state is shared
+        // across all routes within application
+        let addr = ctx.address();
+        self.addr
+            .send(websocket::Connect {
+                addr: addr.recipient(),
+            })
+            .into_actor(self)
```

Comment:
Does actix still not allow Handlers and lifecycle methods to be async blocks? :grimacing:

Reply:
Does not look like it.

Reply:
Very sad